### PR TITLE
CPU-Strided-Complex Fixes for real and imag ops

### PR DESCRIPTION
**Pull request by dylanbespalko**
In-tree changes to pytorch to support complex numbers are being submitted here.
Out-of-tree support for complex numbers is here: [pytorch-cpu-strided-complex extension](https://gitlab.com/pytorch-complex/pytorch-cpu-strided-complex)

- [x]  Replaced std:real(a) with a.real() in kernel level code.
- [x]  Fixed Vec256_base implementation of complex ops so that it works correctly on Non-AVX devices. 
- [x]  Fix NumericUtils.h 

cc: @iotamudelta, @ezyang, @bddppq, @zasdfgbnm

### CODE REVIEW

**Comment by dylanbespalko** [comment]
@iotamudelta,

After these changes, there are no more references to `std::real()` or `std::imag()` in the code.

**Comment by ezyang** [comment]
@zasdfgbnm @iotamudelta I'm going to defer review to you, thanks!

**Review by iotamudelta — APPROVED** [review]
Thanks! Should be fine. If we have any internal fallout, I'll fix.

File: aten/src/ATen/cpu/vec256/vec256_base.h [hunk 1]
```diff
@@ -179,39 +179,75 @@ struct Vec256 {
...
     // Specifically map() does not perform the type conversion needed by abs.
-    return map([](T x) { return (T)std::abs(x); });
+    return map([](T x) { return static_cast<T>(std::abs(x)); });
   }
+  template <typename other_t_angle = T,
+            typename std::enable_if<!c10::is_complex_t<other_t_angle>::value, int>::type = 0>
   Vec256<T> angle() const {
+    // other_t_angle is for SFINAE and clarity. Make sure it is not changed.
+    static_assert(std::is_same<other_t_angle, T>::value, "other_t_angle must be T");
     return *this;
```
Comment (zasdfgbnm) [inline aten/src/ATen/cpu/vec256/vec256_base.h]: Isn't the angle for real numbers always zero?

File: aten/src/ATen/cpu/vec256/vec256_base.h [hunk 2]
```diff
@@ -179,39 +179,75 @@ struct Vec256 {
...
+    static_assert(std::is_same<complex_t_real, T>::value, "complex_t_real must be T");
+    // Specifically map() does not perform the type conversion needed by abs.
+    return map([](T x) { return static_cast<T>(x.real()); });
+  }
+  template <typename other_t_imag = T,
+            typename std::enable_if<!c10::is_complex_t<other_t_imag>::value, int>::type = 0>
   Vec256<T> imag() const {
+    // other_t_imag is for SFINAE and clarity. Make sure it is not changed.
+    static_assert(std::is_same<other_t_imag, T>::value, "other_t_imag must be T");
     return *this;
```
Comment (zasdfgbnm) [inline aten/src/ATen/cpu/vec256/vec256_base.h]: Isn't the imag of a real number always 0?

**Comment by zasdfgbnm** [comment]
The rest LGTM

File: aten/src/ATen/cpu/vec256/vec256_base.h (hunk 2 above)
Comment (dylanbespalko) [reply aten/src/ATen/cpu/vec256/vec256_base.h]: Fixed imag, angle, and conj for non-complex numbers. Thanks.

**Review by zasdfgbnm — APPROVED** [review]
LGTM

**Comment by dylanbespalko** [comment]
@ezyang,

I think this one is ready. Take care.

**Pull request closed without merging**